### ISSUE TITLE
Retry an endpoint if it fails due to a reddit server error (HTTP 5xx)

### ISSUE THREAD

**Issue by trevorsenior** [trevorsenior/snoocore#82]
When setting up Travis CI some of the endpoints will fail if Reddit is busy:

https://travis-ci.org/trevorsenior/snoocore/builds/47476423

We should probably have some sort of feature that will retry the endpoint in these cases, and after a set number of retries fail (can be adjusted, or removed all together in the configuration)

- - -

Update:

We have to make sure that Reddit actually responds with a HTTP 5xx when the servers are overloaded, but this HTML output makes me think that it does:

```html
<h2>all of our servers are busy right now</h2><h3>please try again in a minute</h3>
<p>(error code: 503) <!-- pretty sure it's a HTTP 5xx from this output --></p>
```

We should handle all HTTP 5xx errors the same way and retry reddit "x" amount of times before giving up. 

- - -

As for implementation / testing this we can follow exactly what damianb said (https://github.com/trevorsenior/snoocore/issues/82#issuecomment-70701555) having a local server running that can respond with HTTP 5xx. 

We already spawn a local server for headless authentication purposes so this will work with Travis-CI. As for specifying it to hit the local server vs. reddit, it's possible to adjust the [(hidden) variables](https://github.com/trevorsenior/snoocore/blob/2.3.2/Snoocore.js#L36) in Snoocore to accommodate this for the calls we are testing.

**Comment by trevorsenior** [comment]
Testing this is tricky, as it's not something that is reproducible. Will probably need to look into testing against a local copy of reddit vs. hitting the website for the test cases. This may not be something that can be tested with Travis-CI however.

**Comment by damianb** [comment]
Not tricky at all. Aim JSON request against a locally spawned server, have it respond with HTTP 500. It should try again, which should hit a second codepath in the locally spawned http server.

If the retry is skipped (there will have to be a post-test cleanup), the server will likely still be spawned and can be killed (you're looking at probably having to use a more global variable to test this with), and the test is marked as a failure.

**Comment by trevorsenior** [comment]
Thanks, I've updated the main comment & updated the issue title. I didn't even consider making it more broad to handle all HTTP 5xx errors, but it makes more sense.

**Comment by trevorsenior** [comment]
Going to add in a new event that lets the user of the library know when we have hit a 5xx error. 

```javascript
reddit.on('5xx_error', function() {
  // do something
});
```

Useful when we want to notify the user why the request is taking longer than usual.

**Comment by damianb** [comment]
that would probably be useful - right now i'm having to trap errors through something similar.

be sure to provide the Error instance, the 50x status code may be important to downstream.

**Comment by go1dfish** [comment]
This is how I accomplished conditional retry logic in Nodewhal:

https://github.com/go1dfish/nodewhal/blob/master/schedule.js#L15

https://github.com/go1dfish/nodewhal/blob/master/nodewhal.js#L335

But I like the idea of error events as well, especially if they provide a callback that allows the determination of whether or not to retry.

**Comment by trevorsenior** [comment]
@go1dfish This will work the same way as the Nodewhal implementation in that it'll rely on an initial configuration value for the the number of retries. Something like:

```javascript
new Snoocore({ retryAttempts: 10 });
```

If it is not set, it will default to x number of retries.

> provide a callback that allows the determination of whether or not to retry.

This is an interesting idea. What use case did you have in mind for this?

**Comment by go1dfish** [comment]
What if the callback just returned the number of tries left to execute (and this was passed into the callback as well).

then your callback can look like this:


    function(error, ... maybe other stuff..., triesLeft) {
      if (typeof triesLeft === 'undefined') {
        triesLeft = DEFAULTTRIESLEFT;
      } else if (!triesLeft) {return 0;}
      return triesLeft - 1;
    }
      
That could be a sensible default maybe, to aways retry config.retryAttempts times.

Having the callback return the number of tries left lets developers adjust retry behavior more for different calls/error conditions within the same app.

**Comment by damianb** [comment]
tbh I'd just put that as an extra property of the Error instance being passed. \**shrug**

**Comment by trevorsenior** [comment]
@damianb Would this be too much information for the error callback?

- retry attempts left
- HTTP status (5xx - numeric)
- HTTP Response Body (raw string response body from reddit)
- The endpoint used (will be in [this format](https://github.com/trevorsenior/snooform#data-format-json-example))
- The given args for this the call

**Comment by damianb** [comment]
I saw the information being provided earlier when I'd broke my own code in trainerred - it's a good summary actually.

As long as it all is available via properties of the Error object, and we're able to programatically work with it, I'd say s'all good.

**Comment by go1dfish** [comment]
@trevorsenior That sounds like plenty of data to me.

**Comment by trevorsenior** [comment]
I'll close this issue once the documentation has been updated for it, but this has been released in `2.5.0` and is available on npm. Basically there are two new global options:

```javascript
retryAttempts: 10, // default to ten
retryDelay: 3000 // default to 3 second wait between retry attempts
```

These two options can also be used on a *per call* basis as well.

**Comment by damianb** [comment]
Assuming then that those two retry params are now reserved in call param properties?

Hmm, reading through the doc page, could be made clearer. I'll open another issue.
